File: ios/LowerBackStretching/Share/RoutineShareLink.swift
import Foundation

/// A custom-scheme deep link that round-trips a custom routine across
/// devices: the recipient taps the link or scans the QR code, the OS
/// routes back to the app, and the importer creates a copy of the
/// routine locally.
///
/// Format: `lowerbackstretching://routine?name=<urlenc>&ids=<csv>`
///
/// The link is self-contained — no server is involved — so it works
/// offline and survives the original sender deleting their copy.

let routineLinkScheme = "lowerbackstretching"
let routineLinkHost = "routine"

struct SharedRoutine: Equatable {
    let name: String
    let stretchIds: [String]
}

func buildRoutineLink(name: String, stretchIds: [String]) -> String {
    var components = URLComponents()
    components.scheme = routineLinkScheme
    components.host = routineLinkHost
    components.queryItems = [
        URLQueryItem(name: "name", value: name),
        URLQueryItem(name: "ids", value: stretchIds.joined(separator: ",")),
    ]
    return components.url?.absoluteString ?? ""
}

/// Parse `link` back into a `SharedRoutine`. Returns nil if the link
/// isn't ours, is missing required params, or carries an empty id list.
func parseRoutineLink(_ link: String) -> SharedRoutine? {
    guard let components = URLComponents(string: link) else { return nil }
    guard components.scheme == routineLinkScheme, components.host == routineLinkHost else {
        return nil
    }
    let queryItems = Dictionary(
        uniqueKeysWithValues: (components.queryItems ?? []).map { ($0.name, $0.value ?? "") }
    )
    guard let rawName = queryItems["name"], let rawIds = queryItems["ids"] else { return nil }
    let trimmedName = rawName.trimmingCharacters(in: .whitespacesAndNewlines)
    let ids = rawIds.split(separator: ",").map { $0.trimmingCharacters(in: .whitespaces) }.filter { !$0.isEmpty }
    guard !trimmedName.isEmpty, !ids.isEmpty else { return nil }
    return SharedRoutine(name: trimmedName, stretchIds: ids)
}

File: android/core/src/main/java/com/lowerbackstretching/core/RoutineShareLink.kt
package com.lowerbackstretching.core

import java.net.URI
import java.net.URLDecoder
import java.net.URLEncoder

/**
 * A custom-scheme deep link that round-trips a custom routine across
 * devices: the recipient taps the link or scans the QR code, the OS
 * routes back to the app, and the importer creates a copy of the
 * routine locally.
 *
 * Format: `lowerbackstretching://routine?name=<urlenc>&ids=<csv>`
 *
 * The link is self-contained — no server is involved — so it works
 * offline and survives the original sender deleting their copy.
 */

const val ROUTINE_LINK_SCHEME = "lowerbackstretching"
const val ROUTINE_LINK_HOST = "routine"

data class SharedRoutine(val name: String, val stretchIds: List<String>)

fun buildRoutineLink(name: String, stretchIds: List<String>): String {
    val encName = URLEncoder.encode(name, Charsets.UTF_8.name())
    val ids = stretchIds.joinToString(",")
    val encIds = URLEncoder.encode(ids, Charsets.UTF_8.name())
    return "$ROUTINE_LINK_SCHEME://$ROUTINE_LINK_HOST?name=$encName&ids=$encIds"
}

/**
 * Parse [link] back into a [SharedRoutine]. Returns null if the link
 * isn't ours, is missing required params, or carries an empty id list.
 *
 * Uses java.net.URI so this is callable from JVM unit tests, not just
 * instrumented tests.
 */
fun parseRoutineLink(link: String): SharedRoutine? {
    val uri = runCatching { URI(link) }.getOrNull() ?: return null
    if (uri.scheme != ROUTINE_LINK_SCHEME || uri.host != ROUTINE_LINK_HOST) return null
    val params = parseQuery(uri.rawQuery ?: return null)
    val name = params["name"]?.let { decodeOrNull(it) } ?: return null
    val ids = params["ids"]?.let { decodeOrNull(it) }
        ?.split(",")?.map { it.trim() }?.filter { it.isNotEmpty() }
        ?: return null
    if (ids.isEmpty() || name.isBlank()) return null
    return SharedRoutine(name = name.trim(), stretchIds = ids)
}

private fun parseQuery(raw: String): Map<String, String> =
    raw.split("&").mapNotNull { pair ->
        val eq = pair.indexOf('=')
        if (eq <= 0) null else pair.substring(0, eq) to pair.substring(eq + 1)
    }.toMap()

private fun decodeOrNull(s: String): String? =
    runCatching { URLDecoder.decode(s, Charsets.UTF_8.name()) }.getOrNull()
